text/0000-modular-queries.md

```diff
@@ -0,0 +1,274 @@
+
+
+- Feature Name: rfc-0000-modular-queries
+- Start Date: (fill me in with today's date, 2021-07-24)
+- Tremor Issue: [tremor-rs/tremor-runtime#940](https://github.com/tremor-rs/tremor-runtime/issues/940)
+- RFC PR: [tremor-rs/tremor-rfcs#TODO](https://github.com/tremor-rs/tremor-rfcs/pull/TODO)
+
+
+# Summary
+
+Add support for modular queries to Tremor's query language, Trickle, so that distinct subgraphs could be reused and composed into higher level queries.
+
+# Motivation
+
+Subqueries would allow composition of smaller, reusable queries into higher level queries.
+
+# Guide-level explanation
+
+## Definition
+
+![DefineSubqueryDefn](../img/rfcs/0000-modular-queries/DefineSubqueryDefn.png)
+
+```
+DefineSubqueryDefn ::= DocComment? 'define' 'subquery' Id ('from' Ports)? ('into' Ports)? WithPartialParams? Subquery
+```
+
+```
+## Documentation Comment
+define subquery custom_subquery
+from input_stream_1, input_stream_2
+into output_stream
+with
+ param1 = "foo",
+ param2 = 42
+subquery
+ select event from input_stream_1 where event.name == args.param1 into output_stream;
+ select event from input_stream_2 where event.id == args.param2 into output_stream;
+end;
+```
+
+- The `with` clause can be used to pass in parameters when required.
+    - The values given to the parameters in the with clause here act as their default values.
+    - Parameters are accessible through `args` inside the subquery.
+- The `from` clause can be used to define input streams.
+    - If elided, `in` stream is attached.
```

Comment:
Do i understand this correctly?
If 1 or more ports are provided in `from`, only these streams are created? No `in` stream? So, if I use `from` i need to list `in` as well if i want to use it?

Reply:
Yes.
The user needs to be explicit about the ports _except_ for when they are **only** using the `in` port. If they add an additional port, they would need to then need to explicitly include `in` as well.